go/private/extensions.bzl

```diff
@@ -15,40 +15,130 @@ _download_tag = tag_class(
+                goos = download_tag.goos,
+                goarch = download_tag.goarch,
+                sdk_repo = name,
+                sdk_type = "remote",
+                sdk_version = download_tag.version,
+            ))
+
+        for index, host_tag in enumerate(module.tags.host):
             # Dependencies can rely on rules_go providing a default remote SDK. They can also
             # configure a specific version of the SDK to use. However, they should not add a
             # dependency on the host's Go SDK.
             if not module.is_root:
                 fail("go_sdk.host: cannot be used in non-root module " + module.name)
-            go_host_sdk(
-                name = host_tag.name,
+
+            name = host_tag.name or _default_go_sdk_name(
+                module = module,
+                multi_version = multi_version_module[module.name],
+                tag_type = "host",
+                index = index,
+            )
+            go_host_sdk_rule(
+                name = name,
                 version = host_tag.version,
-                register_toolchains = False,
             )
 
+            toolchains.append(struct(
+                goos = "",
+                goarch = "",
+                sdk_repo = name,
+                sdk_type = "host",
+                sdk_version = host_tag.version,
+            ))
+
+    if len(toolchains) > _MAX_NUM_TOOLCHAINS:
+        fail("more than {} go_sdk tags are not supported".format(_MAX_NUM_TOOLCHAINS))
+
+    # Toolchains in a BUILD file are registered in the order given by name, not in the order they
+    # are declared:
+    # https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/packages/Package.java;drc=8e41dce65b97a3d466d6b1e65005abc52a07b90b;l=156
+    # We pad with an index that lexicographically sorts in the same order as if these toolchains
+    # were registered using register_toolchains in their MODULE.bazel files.
+    go_multiple_toolchains(
+        name = "go_toolchains",
+        prefixes = [
+            _toolchain_prefix(index, toolchain.sdk_repo)
+            for index, toolchain in enumerate(toolchains)
+        ],
+        geese = [toolchain.goos for toolchain in toolchains],
+        goarchs = [toolchain.goarch for toolchain in toolchains],
+        sdk_repos = [toolchain.sdk_repo for toolchain in toolchains],
+        sdk_types = [toolchain.sdk_type for toolchain in toolchains],
+        sdk_versions = [toolchain.sdk_version for toolchain in toolchains],
+    )
+
+def _default_go_sdk_name(*, module, multi_version, tag_type, index):
+    # Keep the version out of the repository name if possible to prevent unnecessary rebuilds when
+    # it changes.
+    return "{name}_{version}_{tag_type}_{index}".format(
+        name = module.name,
+        version = module.version if multi_version else "",
+        tag_type = tag_type,
+        index = index,
+    )
+
+def _toolchain_prefix(index, name):
```

Comment:
It might be helpful to have a comment giving an example of this string transformation, as it is a bit difficult to understand.